### PR TITLE
HTML API: Add set_modifiable_text() for replacing text nodes.

### PR DESCRIPTION
**Pull request by dmsnell**
Trac ticket: Core-61617

This patch introduces a new method, `set_modifiable_text()` to the Tag Processor, which makes it possible and safe to replace text nodes within an HTML document, performing the appropriate escaping.

This method can be used in conjunction with other code to modify the text content of a document, and can be used for transforming HTML in a streaming fashion.

### CODE REVIEW

File: tests/phpunit/tests/html-api/wpHtmlTagProcessorModifiableText.php [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+			"Should have transformed the HTML as expected why modifying the target node's modifiable text."
+		);
+	}
+
+	/**
+	 * Data provider.
+	 *
+	 * @return array[]
+	 */
+	private static function data_tokens_with_basic_modifiable_text_updates() {
```
Comment (gziolo) [inline tests/phpunit/tests/html-api/wpHtmlTagProcessorModifiableText.php]: There are some CI failures:

Screenshot 2024-07-12 at 09 36 20

Does it have to be a public method maybe?

**Review by gziolo — COMMENTED** [review]
That doesn't look that complicated to add support for replacing text nodes. This is very exciting!

There are some edge cases covered that I don't have enough insights to review quickly, but I'm looking forward to seeing these changed landed in core.

File: tests/phpunit/tests/html-api/wpHtmlTagProcessorModifiableText.php (hunk 1 above)
Comment (dmsnell) [reply tests/phpunit/tests/html-api/wpHtmlTagProcessorModifiableText.php]: yeah I've done this before on accident. made the data provider public in 0340ee3

File: src/wp-includes/html-api/class-wp-html-tag-processor.php [hunk 2]
```diff
@@ -2881,6 +2881,157 @@ public function get_modifiable_text() {
...
+				 * This is over-protective, but ensures the update doesn't break
+				 * out of the SCRIPT element. A more thorough check would need to
+				 * ensure that the script closing tag doesn't exist, and isn't
+				 * also "hidden" inside the script double-escaped state.
+				 *
+				 * It may seem like replacing `</script` with `<\/script` would
+				 * properly escape these things, but this could mask regex patterns
+				 * that previously worked. Resolve this by not sending `</script`
+				 */
+				if ( false !== stripos( $plaintext_content, '</script' ) ) {
```
Comment (gziolo) [inline src/wp-includes/html-api/class-wp-html-tag-processor.php:3053]: I can’t locate a unit test covering this edge case. Is it included?

**Review by gziolo — APPROVED** [review]
It all looks great to me.

**Comment by adamziel** [comment]
`get_updated_html()` might need this adjustment:

```
		if ($this->get_token_type() === '#tag') {
			$this->bytes_already_parsed = $before_current_tag;
		}
```

Without it, replacing a longer string with a shorter one can make `$this->bytes_already_parsed` negative.

File: src/wp-includes/html-api/class-wp-html-tag-processor.php (hunk 2 above)
Comment (dmsnell) [reply src/wp-includes/html-api/class-wp-html-tag-processor.php:3053]: Added in [944c8cb](https://github.com/WordPress/wordpress-develop/pull/7007/commits/944c8cb31fa0fc8b3794ce59935e001d3bebe403)

**Comment by dmsnell** [comment]
> get_updated_html() might need this adjustment:

Thanks @adamziel - this level of review is so helpful! Unfortunately, when I added a test to try and trigger the condition, I couldn't. What do you think? Do you see any problem with this?

https://github.com/WordPress/wordpress-develop/pull/7007/files#diff-ef817e9b61dcdc18d5e1a84bfbb55370916b179cfed01eab03121456d7ce847cR42-R86

**Comment by dmsnell** [comment]
Merged in [[58829]](https://core.trac.wordpress.org/changeset/58829)
https://github.com/wordpress/wordpress-develop/commit/c6aaf0a7db6e0555ac61b96cecb699b850fb3187

**Comment by adamziel** [comment]
> Thanks @adamziel - this level of review is so helpful! Unfortunately, when I added a test to try and trigger the condition, I couldn't. What do you think? Do you see any problem with this?

That test looks good. I'd also check for regular tags like `DIV` and then for HTML comments. I ran into that in https://github.com/adamziel/site-transfer-protocol, although now can't find the commit where I did it. Eventually I've switched to [a different method of setting modifiable text](https://github.com/adamziel/site-transfer-protocol/blob/3486d676a5fa2a76b719e3e6159a80e3013bed8a/src/WP_Block_Markup_Processor.php#L285), and now I'm happy to migrate to this now that it's merged. Lovely work and thank you!

**Comment by dmsnell** [comment]
@adamziel if you ever find this problem again I'll be happy to fix it ASAP

**Comment by adamziel** [comment]
> Thanks @adamziel - this level of review is so helpful! Unfortunately, when I added a test to try and trigger the condition, I couldn't. What do you think? Do you see any problem with this?

I just isolated the problem @dmsnell:

```php
$p = new WP_HTML_Tag_Processor('Hello there');
$p->next_token();
$p->set_modifiable_text('Short');
echo $p->get_updated_html();
```

It's not occurring with this patch:

```diff
-		$this->bytes_already_parsed = $before_current_tag;
+		if ($this->get_token_type() === '#tag') {
+			$this->bytes_already_parsed = $before_current_tag;
+		}
```

**Comment by gziolo** [comment]
@adamziel, thank you for the report. I was able to create 3 failing tests where it doesn't work as expected:

```
There were 3 failures:

1) Tests_HtmlApi_WpHtmlTagProcessorModifiableText::test_get_modifiable_text_is_consistent_after_writes_when_text_shorter
Should have found updated text.
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'shorter text'
+'xt'

/var/www/tests/phpunit/tests/html-api/wpHtmlTagProcessorModifiableText.php:116
phpvfscomposer:///var/www/vendor/phpunit/phpunit/phpunit:106

2) Tests_HtmlApi_WpHtmlTagProcessorModifiableText::test_get_modifiable_text_is_consistent_after_writes_when_text_longer
Should have found updated text.
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'a bit longer text'
+'it longer text'

/var/www/tests/phpunit/tests/html-api/wpHtmlTagProcessorModifiableText.php:155
phpvfscomposer:///var/www/vendor/phpunit/phpunit/phpunit:106

3) Tests_HtmlApi_WpHtmlTagProcessorModifiableText::test_get_modifiable_text_is_consistent_after_writes_when_text_after_closed_tag_element
Should have found updated text.
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'<p>some content</p>a bit longer text'
+'it longer text'

/var/www/tests/phpunit/tests/html-api/wpHtmlTagProcessorModifiableText.php:197
phpvfscomposer:///var/www/vendor/phpunit/phpunit/phpunit:106
```

I will work on the patch and improve the tests I quickly drafted:
- https://github.com/WordPress/wordpress-develop/pull/7577
